src/frame/mod.rs

```diff
@@ -1,115 +1,82 @@
-        // read flag bytes
-
-        let flg_byte = input[0];
-        assert!((flg_byte & 0b11000000) as usize == 1); // version is always 01
-
-        let block_mode = if (flg_byte & 1 << 5) as usize == 1 {
-            BlockMode::Independent
-        } else {
-            BlockMode::Linked
-        };
-        let content_checksum_flag = if (flg_byte & 1 << 2) as usize == 1 {
-            Checksum::ChecksumEnabled
-        } else {
-            Checksum::NoChecksum
-        };
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
 
-        // let content_size_included = (flg_byte & 1 << 4) == 1;
+impl std::error::Error for Error {}
 
-        LZ4FFrameInfo {
-            block_mode,
-            content_checksum_flag,
-            block_size_id: BlockSize::Default, // TODO
-            content_size: None,                // TODO
-        }
-    }
+/// Compress all bytes of `input`.
+pub fn compress(input: &[u8]) -> Vec<u8> {
+    compress_with(FrameInfo::default(), input)
+}
 
-    /// writes flag byte
-    ///
-    /// FLG byte
-    /// BitNumber   7-6         5                      4               3             2                 1           0
-    /// FieldName   Version     Block Independence     Block-Checksum  Content-Size  Content-Checksum  Reserved    DictID
-    ///
-    /// Content-Size (Optional) - The size of the uncompressed data included within the frame will be present as an 8 bytes unsigned little endian value, after the flags
-    fn write_flg_byte(&self) -> u32 {
-        let version = 1 << 6; // always 01
-        let mut res = version;
-        res |= (self.block_mode as u32) << 5;
-        if self.content_size.is_some() {
-            res |= 1 << 3; // set the content_size flag
-        }
-        res |= (self.content_checksum_flag as u32) << 2;
-        // res |= 0 << 4;
-        res
-    }
+/// Compress all bytes of `input`.
+pub fn compress_with(frame_info: FrameInfo, input: &[u8]) -> Vec<u8> {
+    let buffer = Vec::with_capacity(
+        header::MAX_FRAME_INFO_SIZE
+            + header::BLOCK_INFO_SIZE
+            + crate::block::compress::get_maximum_output_size(input.len()),
+    );
+    let mut enc = FrameEncoder::with_frame_info(frame_info, buffer);
+    enc.write_all(input).unwrap();
+    enc.finish().unwrap()
+}
 
-    /// writes block descriptor byte
-    /// Block Maximum Size
-    ///
-    /// This information is useful to help the decoder allocate memory. Size here refers to the original (uncompressed) data size.
-    /// Block Maximum Size is one value among the following table :
-    /// 0       1       2       3       4       5       6       7
-    /// N/A     N/A     N/A     N/A     64 KB   256 KB  1 MB    4 MB
-    fn write_bd_byte(&self) -> u32 {
-        1 << (self.block_size_id as u32)
-    }
+/// Decompress all bytes of `input` into a new vec.
+pub fn decompress(input: &[u8]) -> Result<Vec<u8>, Error> {
```

Comment:
data can be very large, so we shouldn't require it to be in memory and accept `impl Read` instead

```pub fn decompress(input: &[u8]) -> Result<Vec<u8>, Error> {```

not sure about the api implications downstream here though..

Reply:
This is supposed to be an utility function to mirror the ones in the block module, in case the user wants to work with bytes/Vec.
Users wanting to interface with io::{Read, Write} are better of with the equivalent FrameEncoder/Decoder.

Reply:
The decision for the block API to support only in memory data is purely for performance reasons, `Read` and `Write` as an API would be better

Reply:
Indeed :thinking:  These utility functions should probably take io::{Read, Write} whenever possible.

Reply:
There's benefit in having easily reachable versions for slice/Vec as they doesn't require using `io::copy` (which involves an intermediary buffer) and also allow better preallocation.

For that reason I added `compress_into`, `compress_into_with` and `decompress_into` which take `io::{Read,Write}` for input/output respectively. It's not worth providing all combinations (eg. input is a slice output is an io::Write), so this probably strikes a good trade-off.

What do you think?

Reply:
I think this depends on the performance gain, you can also preallocate the vec and pass it as the `impl Write`

Reply:
Right, a cautions caller could preallocate, which leaves only io::copy overhead.
Don't you think this is the most common usage, so it warrants the optimized functions?

Reply:
since a Vec implements `Write` the usage would be same, so there needs to be some performance gain to justify the method

Reply:
I ran **compression** benchmarks and both the io::copy and not preallocating overheads are negligible for small data (<=64kb). Probably because compression is much slower than memory copies.

I only measured significant overhead with the 10mb dataset. 2% overhead for the io::copy and 2% more if the Vec is not preallocated.

For **decompression** I measured 3% overhead by not preallocating the output.

Surprisingly (to me at least) the io::copy version (decompress_into) was initially a whopping 18% faster than the specialized version (decompress) which uses read_to_end. This is because the default io::Read::read_to_end implementation is really suboptimal in this case.

Fixing FrameDecoder::read_to_end makes it 5% faster than the equivalent io::copy version.

At least we know no version is awfully worse than the other, and than pre-allocation doesn't matter much except for large data.

Reply:
With that said I think the current functions in frame/mod.rs are reasonable as they strike a good balance between API surface, ergonomics and optimization.

Reply:
Thank you for the insight. With gains of 2-5%, I'm not really convinced for the reduced API ergonomics. With frames, the default case is to stream them I think. In most cases, e.g. a web server, allocating 10MB for a decompression could already be too much. If you have always smaller data, you can use the faster block API.

I had a look at the snap crate: https://docs.rs/snap/1.0.5/snap/index.html

1. There are no such utility functions
2. There is `write::FrameEncoder` `read::FrameDecoder` and (surprisingly)`read::FrameEncoder ` 

I think the api of the utility function is easier to grasp, than `FrameEncoder` and `FrameDecoder`, but maybe that's just me and this can also be completely covered by documentation. 
Generally they are not necessary at all. So my feeling is, it's better to have a minimal, powerful API like the snap crate and document it well for all cases.

Reply:
Seems sensible. I updated the PR as such.